spec.bs

```diff
@@ -980,6 +980,77 @@ To <dfn>add a part</dfn> given a [=pattern parser=] |parser|, a string |prefix|,
+    1. If |part|'s [=part/type=] is "<a for=part/type>`segment-wildcard`</a>", then set |regexp value| to the result of running [=generate a segment wildcard regexp=] give |options|.
+    1. Else if |part|'s [=part/type=] is "<a for=part/type>`full-wildcard`</a>", then set |regexp value| to [=full wildcard regexp value=].
+    1. If |part|'s [=part/prefix=] is the empty string and |part|'s [=part/suffix=] is the empty string:
+      <div class=note>
+      <p>If there is no [=part/prefix=] or [=part/suffix=] then we generate a simple capturing group.  It uses the form:
+      <p>`(<regexp value>)<modifier>`.
+      </div>
+      1. Append "`(`" to the end of |result|.
+      1. Append |regexp value| to the end of |result|.
+      1. Append "`)`" to the end of |result|.
+      1. Append the result of running [=convert a modifier to a string=] give |part|'s [=part/modifier=] to the end of |result|.
+      1. [=Continue=].
+    1. If |part|'s [=part/modifier=] is "<a for=part/modifier>`none`</a>" or "<a for=part/modifier>`optional`</a>":
+      <div class=note>
+      <p>This section handles non-repeating parts with a [=part/prefix=] and/or [=part/suffix=].  There is an inner capturing group that contains the primary |regexp value|.  The inner group is then combined with the [=part/prefix=] and/or [=part/suffix=] in an outer non-capturing group.  Finally the modifier is applied.  The overall form is:
+      <p>`(?:<prefix>(<regexp value>)<suffix>)<modifier>`.
+      </div>
+      1. Append "`(?:`" to the end of |result|.
+      1. Append the result of running [=escape a regexp string=] given |part|'s [=part/prefix=] to the end of |result|.
+      1. Append "`(`" to the end of |result|.
+      1. Append |regexp value| to the end of |result|.
+      1. Append "`)`" to the end of |result|.
+      1. Append the result of running [=escape a regexp string=] given |part|'s [=part/suffix=] to the end of |result|.
+      1. Append "`)`" to the end of |result|.
+      1. Append the result of running [=convert a modifier to a string=] give |part|'s [=part/modifier=] to the end of |result|.
+      1. [=Continue=].
+    1. [=Assert=]: |part|'s [=part/modifier=] is "<a for=part/modifier>`zero-or-more`</a>" or "<a for=part/modifier>`one-or-more`</a>".
+    1. [=Assert=]: |part|'s [=part/prefix=] is not the empty string or |part|'s [=part/suffix=] is not the empty string.
+      <div class=note>
+      <p class=allow-2119>Repeating parts with a [=part/prefix=] and/or [=part/suffix=] are dramatically more complicated.  We want to exclude the initial [=part/prefix=] and the final [=part/suffix=], but include them between any repeated elements.  To achieve this we provide a separate initial expression that excludes the [=part/prefix=].  Then the expression is duplicated with the [=part/prefix=]/[=part/suffix=] values included in an optional repeating element.  If zero values are permitted then a final optional modifier may be appended.  The overall form:
```

Comment:
```suggestion
      <p class=allow-2119>Repeating parts with a [=part/prefix=] and/or [=part/suffix=] are dramatically more complicated.  We want to exclude the initial [=part/prefix=] and the final [=part/suffix=], but include them between any repeated elements.  To achieve this we provide a separate initial expression that excludes the [=part/prefix=].  Then the expression is duplicated with the [=part/prefix=]/[=part/suffix=] values included in an optional repeating element.  If zero values are permitted then a final optional modifier may be appended.  The overall form is:
```

Reply:
Refactors this to "It uses the following form."